Disable OpenBLAS multithreading

nilsvu opened the pull request:
## Proposed changes

OpenBLAS multithreading conflicts with Charm++ parallelism, so this PR disables it in the Docker container. Also adds a note to the installation instructions to disable the option.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu commented:
@geoffrey4444 Once the flag is set in the Docker container and you have updated your containers, you can remove this line from your Ocean submit scripts:

```
export OPENBLAS_NUM_THREADS=1
```

I didn't remove the line in this PR so the submit scripts keep working with existing Docker/Singularity containers.

wthrowe commented:
Could we use the c interface (`openblas_set_num_threads(1)`) so we don't have to update the setup on every machine?

nilsvu commented:
Where would be a good place to call this function?

wthrowe commented:
I think either in Main or right before the BLAS calls.  (Looking at the source, calling the function repeatedly with the same argument should be pretty cheap.)

It also presumably needs to be conditional on using OpenBLAS as the BLAS implementation, but CMake should already know if that's true.

nilsdeppe commented:
Does it need to be called on every core/node? If so, then it'll need to be in one of the proc/node init functions.

wthrowe commented:
Good question.  I can't figure that out, so I've filed xianyi/OpenBlas#2809 for clarification.

In the meantime, let's just do `setenv("OPENBLAS_NUM_THREADS", "1", 1)`.  That is definitely global to the entire process (but not thread-safe).  (And then we don't even need tests in CMake.)

nilsdeppe commented:
That sounds good to me! :)

nilsdeppe commented:
LGTM. I'm fine with you squashing once @wthrowe has had a chance to look :)

nilsdeppe commented:
I'm not entire sure that we shouldn't be setting this in an initenode function, though... That seems safer since it'll get run on each MPI rank/process. @wthrowe do you know how `setenv` interacts with MPI as to whether or not it is necessary to do it on every MPI process?

wthrowe commented:
Oh, right, `Main != main`.  Yes, it should be in initnode.

nilsvu commented:
@nilsdeppe @wthrowe would one of you like to take this over? I feel like that could be faster

nilsdeppe commented:
I don't want to, no. I think you should test that what you're changing actually works on 2 nodes, and with both 1 Charm++ node per node and 2 Charm++ nodes per node. We are only voicing concerns and you should test your code to make sure it works correctly.

wthrowe commented:
I don't even have a test case for this, so I don't think taking it over makes sense.

The OpenBLAS people replied to my issue, but I don't understand their response: https://github.com/xianyi/OpenBLAS/issues/2809#issuecomment-685502307

nilsvu commented:
@wthrowe I also don't understand if they recommend to set the env variable or call the C function. I'll just have to do a few tests on Minerva.

nilsdeppe commented:
The best I can tell from their response is "No, it's not thread safe" because of the "tame for thread...", but I don't have any real confidence in that. I think testing on Minerva is the right way to go about it :)

nilsvu commented:
@nilsdeppe @wthrowe here's what I found:

- `setenv("OPENBLAS_NUM_THREADS", "1", 1)` seems to have no effect. I found some people reported similar findings, that setting the env variable within the program has no effect.
- Calling `openblas_set_num_threads(1)` on every node works fine. But this requires us to check if we're using OpenBLAS or another vendor. I couldn't find a canonical way to do this check, either within CMake or in C++. Can you think of a robust way to check for OpenBLAS?

Here's my current suggestions for handling the issue:
1. Try to implement a check for OpenBLAS in `SetupBLAS.cmake`. I don't think we can make that check sufficiently reliable though, because OpenBLAS libraries may be symlinked, not consistently named, etc. This means that if OpenBLAS is being used but we didn't detect it, our run will crash with no indication for the user why it did.
2. Find some way in C++ to check if an OpenBLAS-specific preprocessor macro is defined, or if `openblas_set_num_threads` exists.
3. Rely on the OpenBLAS installation or the runtime environment to disable multithreading (this PR's original approach). If the other approaches are not robust enough, this might be the only reliable option.

What do you think?

nilsvu commented:
Related to option 1.: We could do this:

```cmake
set(BLA_VENDOR OpenBLAS)
find_package(BLAS)
if(BLA_FOUND)
  # We found OpenBLAS, configure appropriately
else()
  set(BLA_VENDOR "")
  find_package(BLAS REQUIRED)
endif()
```

This approach has two flaws that I can see:
- It puts the responsibility to detect the vendor correctly on https://github.com/Kitware/CMake/blob/master/Modules/FindBLAS.cmake, and I'm unsure how reliable that is.
- It will always try to find OpenBLAS first, which might mess with the user's intention.

nilsdeppe commented:
Once you've found BLAS you can do a call into the compiler to see if it can compile with the thread changing function. We do this for detecting warning flags. This can be done in a way that does not require writing a file to disk, which is important to keep CMake running quickly. I don't remember off hand where exactly we do this, but we have such an implementation. This will allow you to robustly detect that the function you need exists

fmahebert commented:
@nilsdeppe Are you referring to `CheckCxxLinkerFlag.cmake`? ~This appears to try compiling from a CMake string.~ We also have `CheckBrokenArray0.cmake` that tries compiling from a file.

Edit: nevermind, `CheckCxxLinkerFlag.cmake` builds a string in CMake but writes it to file before trying compilation.

nilsvu commented:
I think `CheckCxxLinkerFlag.cmake` actually writes to a file, whereas `CheckBrokenArray0.cmake` compiles an existing file from the repository. I'm implementing the latter approach right now.

nilsvu committed:
Support disabling OpenBLAS multithreading

OpenBLAS multithreading conflicts with Charm++ parallelism, so we
need to disable it in parallel executables.

nilsvu commented:
@nilsdeppe @wthrowe Pushed the new version. It works on multiple nodes on Minerva. Could you take another look?

nilsdeppe commented on cmake/SetupBlas.cmake:
```diff
@@ -16,3 +16,20 @@ set_property(
+# Check if we have found OpenBLAS and can disable its multithreading, since it
+# conflicts with Charm++ parallelism. Details:
+# https://github.com/xianyi/OpenBLAS/wiki/Faq#multi-threaded
+try_compile(
```
I specifically requested you don't read a file from disk...

nilsvu replied:
You said "writing a file to disk". I'm not sure this can be done without reading a file from disk (see @fmahebert's and my comment above).

nilsdeppe replied:
Yes, it can be done...

nilsdeppe replied:
See my comment "We do this for warning flags"

nilsvu replied:
This is the code in `cmake/CheckCxxLinkerFlag.cmake`, if that's what you mean...

```cmake
  set(SOURCE_TO_TRY "int main() { return 0; }")
  set(FILE_TO_WRITE
    "${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/CMakeTmp/src.cxx")
  file(WRITE "${FILE_TO_WRITE}" "${SOURCE_TO_TRY}\n")
  set(LINK_COMMAND_TO_TRY
    "${CHARM_COMPILER} ${FLAG_TO_CHECK} <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> -o <TARGET> <LINK_LIBRARIES>")

  if(NOT CMAKE_REQUIRED_QUIET)
    message(STATUS "Performing Test ${FLAG_TO_CHECK}")
  endif()

  try_compile(${FLAG_WORKS}
    ${CMAKE_BINARY_DIR}
    ${FILE_TO_WRITE}
    CMAKE_FLAGS -DCMAKE_CXX_LINK_EXECUTABLE:STRING=${LINK_COMMAND_TO_TRY}
    OUTPUT_VARIABLE OUTPUT)
```

It writes a source file to disk and then reads it again.

nilsdeppe replied:
You're looking at **linker** flags, not compiler flags... `AddCxxFlag.cmake`

nilsvu replied:
I see, that's using `execute_process` instead of `try_compile`. Seems rather painful to do, linking with Blas and everything, when `try_compile` works fine and doesn't notably increase cmake configuration time..

nilsvu replied:
I'll have to look at this again another time, it's late

nilsdeppe replied:
Please use the approach from `AddCxxFlag`. There's a reason we do that: using `try_compile` is very expensive and using it many times really slows down CMake. This is why we do the flags the way they are done. The argument "Using it only once here..." can be applied repeatedly and then CMake takes forever to run. Also keep in mind that, especially on HPC systems, the filesystem can be very slow and while you might not notice a slowdown on your machine with an SSD, that will almost certainly not be true on some network file systems.

nilsdeppe commented on src/Parallel/CharmMain.tpp:
```diff
@@ -111,6 +120,12 @@ void register_init_node_and_proc() noexcept {
+  // Disable OpenBLAS multithreading since it conflicts with Charm++
+  // parallelism. Details:
+  // https://github.com/xianyi/OpenBLAS/wiki/Faq#multi-threaded
+  _registerInitCall(disable_openblas_multithreading, 1);
```
Please move this to a `init_node_func`, as I requested initially.

nilsvu replied:
That's what I thought this is. Could you be more specific?

nilsdeppe replied:
In all the header files for the executables there's an `init_node` vector at the bottom

nilsvu replied:
Sure, are you saying you would prefer adding the `disable_openblas_multithreading` function to _every one_ of those vectors? I added the function in CharmMain.tpp so we don't have to repeat it everywhere.

nilsdeppe replied:
Yes, because it is reasonable to want to disable it

nilsvu committed:
Disable OpenBLAS multithreading in most executables

nilsvu commented:
@nilsdeppe @wthrowe Pushed another version :) please take a look.

wthrowe commented on src/Utilities/Blas.cpp:
```diff
@@ -0,0 +1,18 @@
+namespace {
+extern "C" {
+#ifdef DISABLE_OPENBLAS_MULTITHREADING
+void openblas_set_num_threads(int num_threads);
```
Is there a reason to declare this ourselves instead of including cblas.h?

nilsvu replied:
I added a comment to the code. Our `Blas` library doesn't provide include directories, so `cblas.h` might not be available, or it might be inconsistent with the BLAS library that we're linking in. The reason we don't have the include directories is that CMake's FindBLAS (https://cmake.org/cmake/help/latest/module/FindBLAS.html) doesn't provide it. From what I found they don't bother providing include directories because everyone declares the BLAS functions themselves anyway.

wthrowe commented on src/Utilities/Blas.cpp:
```diff
@@ -0,0 +1,18 @@
+
+#include "Utilities/Blas.hpp"
+
+namespace {
```
I don't think putting an `extern "C"` declaration in an anonymous namespace does anything, and it seems misleading because an anonymous namespace indicates that the contents are local to the file.

nilsvu replied:
removed the anonymous namespace

nilsdeppe reviewed (commented):
I have no additional comments to what @wthrowe has. Please just double check everything works as expected on Minerva before we merge :)
